File: apps/yryvu-app/src/ipc/integrations/issues.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

import type { UserInfo } from "./credentials";
import type { Label } from "./pulls";

/// Cross-provider issue state. None of the supported providers
/// (GitHub / GitLab / Gitea) expose anything beyond open/closed on
/// issues.
export type IssueState = "open" | "closed";

/// Cross-provider issue row payload. Deliberately leaner than
/// `PullRequestSummary` — no merge state, no head/base refs, no
/// review/CI badges.
export interface IssueSummary {
  number: number;
  title: string;
  state: IssueState;
  author: UserInfo;
  createdAt: string;
  updatedAt: string;
  htmlUrl: string;
  labels: Label[];
  assignees: UserInfo[];
  comments: number;
}

/// List issues for `owner/repo` on the named provider. Same auth +
/// error envelope as `integrationListPrs`. The frontend filters
/// providers it can't render via the discriminated resource result;
/// the backend bubbles up a typed `NotImplemented` for the rest.
export function integrationListIssues(
  integrationType: string,
  owner: string,
  repo: string,
): Promise<IssueSummary[]> {
  return invoke<IssueSummary[]>("integration_list_issues", {
    integrationType,
    owner,
    repo,
  });
}

/// Extended issue payload for the detail panel — superset of
/// `IssueSummary` with body markdown + closed_at + milestone.
export interface IssueDetail {
  number: number;
  title: string;
  state: IssueState;
  author: UserInfo;
  createdAt: string;
  updatedAt: string;
  closedAt: string | null;
  htmlUrl: string;
  body: string;
  milestone: string | null;
  labels: Label[];
  assignees: UserInfo[];
  comments: number;
}

/// Fetch a single issue's full detail. Same provider-routing as
/// `integrationListIssues`; the backend returns the cross-provider
/// `IssueDetail` shape.
export function integrationGetIssueDetail(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
): Promise<IssueDetail> {
  return invoke<IssueDetail>("integration_get_issue_detail", {
    integrationType,
    owner,
    repo,
    number,
  });
}

File: crates/yryvu-bridge/src/commands/integrations/issues.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::Path;

use tauri::AppHandle;

use crate::integrations::{self, IssueDetail, IssueSummary};

use super::super::integration_routing::is_self_hosted;
use super::sidecar_path;

/// List issues for `owner/repo` on the named provider. Same auth +
/// hostname path as `integration_list_prs`; supported providers
/// route via [`integrations::list_issues`], others bubble up the
/// typed `NotImplemented` so the frontend can degrade gracefully.
#[tauri::command]
pub async fn integration_list_issues(
    app: AppHandle,
    profile_id: Option<String>,
    integration_type: String,
    owner: String,
    repo: String,
) -> Result<Vec<IssueSummary>, String> {
    let path = sidecar_path(&app)?;
    let auth = tauri::async_runtime::spawn_blocking({
        let integration_type = integration_type.clone();
        let profile_id = profile_id.clone();
        move || {
            integrations::get_integration(&path as &Path, profile_id.as_deref(), &integration_type)
        }
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?
    .ok_or_else(|| format!("integration '{integration_type}' is not connected"))?;
    let hostname = is_self_hosted(&integration_type)
        .then_some(auth.hostname.as_deref())
        .flatten();
    integrations::list_issues(&integration_type, &auth.token, hostname, &owner, &repo)
        .await
        .map_err(|e| e.to_string())
}

/// Fetch one issue's full detail (body + metadata) for the in-app
/// detail panel. Same auth + hostname routing as
/// `integration_list_issues`; supported providers route via
/// [`integrations::get_issue_detail`].
#[tauri::command]
pub async fn integration_get_issue_detail(
    app: AppHandle,
    profile_id: Option<String>,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
) -> Result<IssueDetail, String> {
    let path = sidecar_path(&app)?;
    let auth = tauri::async_runtime::spawn_blocking({
        let integration_type = integration_type.clone();
        let profile_id = profile_id.clone();
        move || {
            integrations::get_integration(&path as &Path, profile_id.as_deref(), &integration_type)
        }
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?
    .ok_or_else(|| format!("integration '{integration_type}' is not connected"))?;
    let hostname = is_self_hosted(&integration_type)
        .then_some(auth.hostname.as_deref())
        .flatten();
    integrations::get_issue_detail(
        &integration_type,
        &auth.token,
        hostname,
        &owner,
        &repo,
        number,
    )
    .await
    .map_err(|e| e.to_string())
}
